Add PromQL Service doc and how to use in Grafana. (#10459)
apache/skywalking: docs/en/api/promql-service.md

# PromQL Service

> PromQL([Prometheus Query Language](https://prometheus.io/docs/prometheus/latest/querying/basics/)) Service
> provide Promql-based expression language and HTTP API.
> Third-party systems or visualization platforms that already support PromQL (such as Grafana),
> If metrics data are required for integration, they can be obtained directly from SkyWalking through the PromQL Service.

Reviewer: ```suggestion
PromQL([Prometheus Query Language](https://prometheus.io/docs/prometheus/latest/querying/basics/)) Service
exposes Prometheus Querying HTTP APIs including the bundled PromQL expression system.
Third-party systems or visualization platforms that already support PromQL (such as Grafana), 
could obtain metrics through PromeQL Service.
```

> Since SkyWalking and Prometheus have different mechanisms for metrics including classification, format, storage, etc.
> The PromQL Service supported will be a subset of the full PromQL.

Reviewer: ```suggestion
As SkyWalking and Prometheus have fundamental differences in metrics classification, format, storage, etc. 
The PromQL Service supported will be a subset of the complete PromQL
```

```

**Note: The operations between vectors require the same metric and labels, and don't support [Vector matching](https://prometheus.io/docs/prometheus/latest/querying/operators/#vector-matching).**

#### [Comparison binary operators](https://prometheus.io/docs/prometheus/latest/querying/operators/#comparison-binary-operators)
| Operator | Definition       | Support |
|----------|------------------|---------|
| ==       | equal            | yes     |
| !=       | not-equal        | yes     |
| \>       | greater-than     | yes     |
| <        | less-than        | yes     |
| \>=      | greater-or-equal | yes     |
| <=       | less-or-equal)   | yes     |
##### Between two scalars
For example:
```shell
1 > bool 2
```

##### Between an instant vector and a scalar
For example:
```shell
service_cpm{service='$service', layer='$layer'} > 1
```

##### Between two instant vectors
For example:
```shell
service_cpm{service='service_A', layer='$layer'} > service_cpm{service='service_B', layer='$layer'}
```

### HTTP API

#### Expression queries

##### [Instant queries](https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries)
```shell
GET|POST /api/v1/query
```

> | Parameter | Definition                                                                                                                          | Support | Optional   |

Reviewer: Could you explain what `Optional` mean? The context seems not very clear.

Author: In the Prometheus official doc, use `Optional` so I think follow this is better.

Reviewer: OK, got it.

- service_traffic
- instance_traffic
- endpoint_traffic

#### [Getting label names](https://prometheus.io/docs/prometheus/latest/querying/api/#getting-label-names)
```shell
GET|POST /api/v1/labels
```

| Parameter | Definition      | Support | Optional |
|-----------|-----------------|---------|----------|
| match[]   | series selector | yes     | yes      |
| start     | start timestamp | **no**  | yes      |
| end       | end timestamp   | **no**  | yes      |

For example:
```shell
/api/v1/labels?match[]=instance_jvm_cpu'
```

Result:
```json
{
    "status": "success",
    "data": [
        "layer",
        "scope",
        "top_n",
        "order",
        "service_instance",
        "parent_service"
    ]
}
```

#### [Querying label values](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-label-values)
```shell
GET /api/v1/label/<label_name>/values
```

| Parameter | Definition      | Support | Optional |
|-----------|-----------------|---------|----------|
| match[]   | series selector | yes     | no       |
| start     | start timestamp | **no**  | yes      |
| end       | end timestamp   | **no**  | yes      |

For example:
```shell

> /api/v1/label/__name__/values'

Reviewer 2: Look like a typo, and there are the same typos in other places, please check

```suggestion
/api/v1/label/__name__/values
```

    ...
  ]
}
```

#### [Querying metric metadata](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-metric-metadata)
```shell
GET /api/v1/metadata
```

| Parameter | Definition                                  | Support | Optional |
|-----------|---------------------------------------------|---------|----------|
| limit     | maximum number of metrics to return         | yes     | **yes**  |
| metric    | **metric name, support regular expression** | yes     | **yes**  |

For example:
```shell
/api/v1/metadata?limit=10'
```

Result:
```json
{
  "status": "success",
  "data": {
    "meter_mysql_instance_qps": [
      {
        "type": "gauge",
        "help": "",
        "unit": ""
      }
    ],
    "meter_apisix_sv_bandwidth_unmatched": [
      {
        "type": "gauge",
        "help": "",
        "unit": ""
      }
    ],
    "service_cpm": [
      {
        "type": "gauge",
        "help": "",
        "unit": ""
      }
    ],
    ...
  }
}
```

## Metrics Type For Query

### Supported Metrics [Scope](../../../oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/query/enumeration/Scope.java)(Catalog)

> Full scopes are not currently supported, please check the following table:

Reviewer: ```suggestion
All scopes are not supported completely, please check the following table:
```

```

- Expression Example:
```shell
service_percentile{service='agent::songs', layer='GENERAL', labels='0,1,2', relabels='P50,P75,P90'}
```

- Result (Instant Query):
```json
{
  "status": "success",
  "data": {
    "resultType": "vector",
    "result": [
      {
        "metric": {
          "__name__": "service_percentile",
          "label": "P50",
          "layer": "GENERAL",
          "scope": "Service",
          "service": "agent::songs"
        },
        "value": [
          1677493380,
          "0"
        ]
      },
      {
        "metric": {
          "__name__": "service_percentile",
          "label": "P75",
          "layer": "GENERAL",
          "scope": "Service",
          "service": "agent::songs"
        },
        "value": [
          1677493380,
          "0"
        ]
      },
      {
        "metric": {
          "__name__": "service_percentile",
          "label": "P90",
          "layer": "GENERAL",
          "scope": "Service",
          "service": "agent::songs"
        },
        "value": [
          1677493380,
          "0"
        ]
      }
    ]
  }
}
```

### Sort Metrics
- Query Labels:
```shell

> --parent_service: Name of the parent service

Reviewer: This should be optional? Do we support that? Such as we have a TopN slow service list.
I know we are not recommending that due to Layers mixed, but in theory, we could do this query if users have only one layer, right?
